packages/utilities/src/classNamesFunction.ts

```diff
@@ -12,29 +25,114 @@ export interface IClassNamesFunctionOptions {
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(
   options: IClassNamesFunctionOptions = {}
 ): (getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined, styleProps?: TStyleProps) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  // We build a trie where each node is a Map. The map entry key represents an argument
+  // value, and the entry value is another node (Map). Each node has a `__retval__`
+  // property which is used to hold the cached response.
+
+  // To derive the response, we can simply ensure the arguments are added or already
+  // exist in the trie. At the last node, if there is a `__retval__` we return that. Otherwise
+  // we call the `getStyles` api to evaluate, cache on the property, and return that.
+  let map: IRecursiveMemoNode = new Map();
+  let resultCount = 0;
 
   const getClassNames = (
     styleFunctionOrObject: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined,
     styleProps: TStyleProps = {} as TStyleProps
   ): IProcessedStyleSet<TStyleSet> => {
-    // styleSet might be undefined if styleFunctionOrObject is undefined, but getStyles should never
-    // ordinarily be undefined (it would hardly make any sense).
-    // However, because we usually use `props.styles` as the argument to an invocation of this method, and
-    // `props.styles` itself is defined as optional, this avoids the need to use `!` at all invocation points.
-    if (styleFunctionOrObject === undefined) {
-      return {} as IProcessedStyleSet<TStyleSet>;
+    let current: Map<any, any> = map;
+    const disableCaching = options.disableCaching;
+
+    if (!options.disableCaching) {
+      current = _traverseMap(map, styleFunctionOrObject as any);
+      current = _traverseMap(current, styleProps);
     }
 
-    const styleSet =
-      styleFunctionOrObject && (typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps!) : styleFunctionOrObject);
+    if (disableCaching || !(current as any)[RetVal]) {
+      if (styleFunctionOrObject === undefined) {
+        (current as any)[RetVal] = {} as IProcessedStyleSet<TStyleSet>;
+      } else if (Array.isArray(styleFunctionOrObject)) {
+        (current as any)[RetVal] = mergeStyleSets(...(styleFunctionOrObject as any).map(_derive, styleProps));
+      } else {
+        (current as any)[RetVal] = mergeStyleSets(
+          typeof styleFunctionOrObject === 'function' ? styleFunctionOrObject(styleProps) : styleFunctionOrObject
+        );
+      }
+
+      if (!disableCaching) {
+        resultCount++;
+      }
+    }
+
+    if (resultCount > MAX_CACHE_COUNT) {
+      map.clear();
+      resultCount = 0;
+
+      // Mutate the options passed in, that's all we can do.
+      options.disableCaching = true;
```

Comment:
How will this get set to false again? Is the goal to permanently disable caching?

Reply:
Yes, it will be "permanently" (current web session) disabled for the current component type. If a component type hits MAX_CACHE_COUNT, it is effectively determined to be uncacheable.